### PR TITLE
fix partition key bug

### LINKED ISSUE #3748: bad partition key when re-keying table
### Hail version:
devel-10a75bb57a6f
### What you did:

```
def test_rekey_correct_partition_key(self):
        ht = hl.utils.range_table(5)
        ht = ht.add_index('a')
        ht = ht.key_by('idx', 'a')
        ht = ht.annotate(b=ht.idx)
        ht = ht.key_by('idx', 'b')
        self.assertEqual(ht.aggregate(agg.sum(ht.idx)), 10)
```

### What went wrong (all error messages here, including the full java stack trace):

```
Java stack trace:
java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:156)
	at is.hail.rvd.OrderedRVDType.<init>(OrderedRVDType.scala:21)
	at is.hail.rvd.OrderedRVDType.copy(OrderedRVDType.scala:121)
	at is.hail.expr.TableKeyBy.execute(Relational.scala:1857)
	at is.hail.expr.TableMapRows.execute(Relational.scala:2090)
	at is.hail.expr.TableKeyBy.execute(Relational.scala:1846)
	at is.hail.expr.ir.Interpret$.apply(Interpret.scala:520)
	at is.hail.expr.ir.Interpret$.apply(Interpret.scala:39)
	at is.hail.table.Table.aggregate(Table.scala:465)
	at is.hail.table.Table.aggregate(Table.scala:446)
	at is.hail.table.Table.aggregateJSON(Table.scala:436)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by jbloom22**
fixes #3748

### CODE REVIEW

File: python/hail/tests/test_api.py [hunk 1]
```diff
@@ -619,6 +619,15 @@ def test_issue_3654(self):
         ht = ht.annotate(x = [1,2])
         self.assertEqual(ht.aggregate(hl.agg.array_sum(ht.x) / [2, 2]), [5.0, 10.0])
 
+    def test_rekey_correct_partition_key(self):
+            ht = hl.utils.range_table(5)
+            ht = ht.add_index('a')
+            ht = ht.key_by('idx', 'a')
+            ht = ht.annotate(b=ht.idx)
+            ht = ht.key_by('idx', 'b')
+            self.assertEqual(ht.aggregate(agg.sum(ht.idx)), 10)
```
Comment (patrick-schultz) [inline python/hail/tests/test_api.py]: Did this fail before the fix? The only way I see to trigger the bug is to `key_by` something between the existing partition key and the existing key. You could load a ((locus), alleles) keyed matrix_table and then key the rows table by locus.

**Comment by jbloom22** [comment]
I assumed it failed on master b/c @maccum just reported it: https://github.com/hail-is/hail/issues/3748

I'll check.

**Comment by maccum** [comment]
Yes, it failed on master. It was failing in OrderedRVDType, because the key `idx` didn't match the partition key `[idx, a]`. The error only happens if `idx` is in both the old and new key.

**Comment by jbloom22** [comment]
I've simplified / improved the test to show both modes of failure that indeed occur on master.

**Comment by patrick-schultz** [comment]
Thanks. I think what's happening is that the IR python generates for `key_by('idx', 'b')` after `key_by('idx', 'a')` has a `key_by('idx')` inserted in between, and that triggers the bug.

**Pull request merged**